src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+                return mb_convert_case($m[0], MB_CASE_TITLE, 'UTF-8');
+            }, $str->string);
+        } else {
+            $firstChar = mb_substr($str->string, 0, 1, 'UTF-8');
+            $str->string = mb_convert_case($firstChar, MB_CASE_TITLE, 'UTF-8').substr($str->string, \strlen($firstChar));
+        }
+
+        return $str;
+    }
+
+    public function toBinary(): BinaryString
+    {
+        return new BinaryString($this->string, $this->locale);
+    }
+
+    public function toGrapheme(): GraphemeString
+    {
+        return new GraphemeString($this->string, $this->locale);
+    }
+
+    public function toUtf8(): Utf8String
+    {
+        return new Utf8String($this->string, $this->locale);
+    }
+
+    /**
+     * @return static
+     */
+    public function trim(string $chars = " \t\n\r\0\x0B\x0C\u{A0}\u{FEFF}"): self
+    {
+        if (!preg_match('//u', $chars)) {
+            throw new InvalidArgumentException('Invalid UTF-8 chars.');
+        }
+        $chars = preg_quote($chars);
+
+        $str = clone $this;
+        $str->string = preg_replace("{^[$chars]++|[$chars]++$}u", '', $str->string);
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    public function trimEnd(string $chars = " \t\n\r\0\x0B\x0C\u{A0}\u{FEFF}"): self
+    {
+        if (!preg_match('//u', $chars)) {
+            throw new InvalidArgumentException('Invalid UTF-8 chars.');
+        }
+        $chars = preg_quote($chars);
+
+        $str = clone $this;
+        $str->string = preg_replace("{[$chars]++$}u", '', $str->string);
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    public function trimStart(string $chars = " \t\n\r\0\x0B\x0C\u{A0}\u{FEFF}"): self
+    {
+        if (!preg_match('//u', $chars)) {
+            throw new InvalidArgumentException('Invalid UTF-8 chars.');
+        }
+        $chars = preg_quote($chars);
+
+        $str = clone $this;
+        $str->string = preg_replace("{^[$chars]++}u", '', $str->string);
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    public function truncate(int $length, string $ellipsis = ''): self
+    {
+        $stringLength = $this->length();
+
+        if ($stringLength <= $length) {
+            return $this;
+        }
+
+        $ellipsisLength = '' !== $ellipsis ? (new static($ellipsis, $this->locale))->length() : 0;
+
+        if ($length < $ellipsisLength) {
+            $ellipsisLength = 0;
+        }
+
+        $str = $this->slice(0, $length - $ellipsisLength);
+
+        return $ellipsisLength ? $str->trimEnd()->append($ellipsis) : $str;
+    }
+
+    /**
+     * @return static
+     */
+    public function upper(): self
+    {
+        $str = clone $this;
+        $str->string = mb_strtoupper($str->string, 'UTF-8');
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    public function withLocale(string $locale): self
```

Comment:
There are multiple places where `string $locale` is used - do you consider making `Locale` class for normalization and validation of the locales? It would simplify many things and make the library more bulletproof. Now you can pass whatever you want as `$locale` which will lead to error, which could be handled better and you wouldn't need to do thing like `$this->locale = $this->withLocale($locale)->locale;` in multiple places too.